source/client/process_impl.cc

```diff
@@ -258,6 +254,7 @@ bool ProcessImpl::run(OutputCollector& collector) {
   try {
     uri.resolve(*dispatcher_, Utility::translateFamilyOptionString(options_.addressFamily()));
   } catch (UriException) {
+    tls_.shutdownGlobalThreading();
```

Comment:
Why is this handled here instead of at the higher level?  Perhaps you want to add some termination function instead and call it here and below.  A function which checks what is running and shuts it all down that can be called in different states.  That might simplify future error handling.

Reply:
I moved the shutdown code to the destructor. 
(Note: Some of the calls we need to do are forgiving when called unpaired; so it's not being done conditionally right now).

Reply:
I have just been debugging code which did complex work in destructors.  Would it be possible to have a shutdown() method  instead?  Sorry, to cause you trouble about this but complex code (e.g. locking, threading) in constructors and destructors have caused me too many sleepless night :)

Reply:
I have had numerous problems with code doing complex things like locking and threads in constructors and destructors... sorry to be a pain, but could you use a shutdown() function instead  and call it before the destructor?

Reply:
Sorry about the dup, github was slow to recognize the comment :(